File: 966-BinarySubarraysWithSum/966-BinarySubarraysWithSum.java
// Last updated: 16/10/2025, 04:25:33
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        int n = nums.length;
        int count = 0;
        HashMap<Integer, Integer> prefSum = new HashMap<>(2);
        int sum = 0;
        prefSum.put(0, 1);
        for(int i = 0; i < n; i++){
            sum += nums[i];
            // if(sum == goal){ // remSum = 0, for this case we already put (0,1) in map
            //     count++;
            // }
            int remSum = sum - goal;
            if(prefSum.containsKey(remSum)){
                count += prefSum.get(remSum);
            }
            prefSum.put(sum, prefSum.getOrDefault(sum, 0) + 1);
        }
        return count;
    }
}

File: 966-BinarySubarraysWithSum/966-BinarySubarraysWithSum.cpp
// Last updated: 16/10/2025, 04:25:37
class Solution {
public:

    //Video: https://www.youtube.com/watch?v=95_sni02RqA
    //Doc: https://leetcode.com/problems/binary-subarrays-with-sum/discuss/2571145/C%2B%2B-Sliding-Window-or-Prefix-Sum-or-Hashing-solution-or-O(N)
    int atMost(vector<int> &nums, int target){
        if(target< 0){ // To handle Goal = 0 case and nums = [0 0 0 0]
            return 0;
        }
       int s = 0, e =0;
       int n = nums.size();
       int sum = 0;
       int count = 0;
        while(e<n){
               sum += nums[e];
            if(sum > target){
                while(sum > target){
                    sum -= nums[s];
                    s++;
                }
            }
            count += e-s+1;
            e++;
      }
        return count;
    }
    
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        
        return atMost(nums,goal) - atMost(nums,goal-1);
        
    }
};